grayscale.py

```diff
@@ -1,336 +1,1055 @@
+        # of the display controller's RC oscillator as well as any timing offsets introduced by the Python code.
+        self.post_frame_cmds = bytearray([0xd3,40+(64-58), 0xa8,58-1])
+
+        # We enhance the greys by modulating the contrast
+        if False:
+            # brightest
+            self.post_frame_adj = [bytearray([0x81,0x5]), bytearray([0x81,0x7f]), bytearray([0x81,0xff])]
+        else:
+            # use setting from thumby.cfg
+            brightnessSetting=2
+            try:
+                with open("thumby.cfg", "r") as fh:
+                    conf = fh.read().split(',')
+                for k in range(len(conf)):
+                    if(conf[k] == "brightness"):
+                        brightnessSetting = int(conf[k+1])
+            except OSError:
+                pass
+            brightnessVals = [0,28,127]
+            brightnessVal = brightnessVals[brightnessSetting]
+            self.post_frame_adj = [bytearray([0x81,brightnessVal>>4]), bytearray([0x81,brightnessVal>>1]), bytearray([0x81,brightnessVal])]
+
+        # On micropython v1.18, it's important to avoid using regular variables for thread sychronisation.
+        # instead, elements of an array/bytearray should be used. We're using a uint32 array here, as that
+        # should hopefully further ensure the atomicity of any element accesses.
+        self._state = array('I', [0,0,0,0xff])
+
+        self.pending_cmds = None
+
+        self.setFont('lib/font5x7.bin', 5, 7, 1)
+        #self.setFont('lib/font8x8.bin', 8, 8, 0)
+
+        self.lastUpdateEnd = 0
+        self.frameRate = 0
+        self.frameTimeMs = 0
+
+        self.fill(Grayscale.BLACK)
+        self._copy_buffers()
+        self.init_display()
+        self.state = Grayscale_ThreadState_Starting
+        _thread.start_new_thread(self._display_thread, ())
+        while self._state[Grayscale_StateIndex_State] != Grayscale_ThreadState_Running:
+            if self._state[Grayscale_StateIndex_State] == Grayscale_ThreadState_Exception:
+                break
+            idle()
+
+
+    @property
+    def error(self):
+        return self._state[Grayscale_StateIndex_State] == Grayscale_ThreadState_Exception
+
+
+    # allow use of 'with'
+    def __enter__(self):
+        return self
+    def __exit__(self, type, value, traceback):
+        self.stop()
+
+
+    def reset(self):
+        self.res(1)
+        utime.sleep_ms(1)
+        self.res(0)
+        utime.sleep_ms(10)
+        self.res(1)
+        utime.sleep_ms(10)
+
+
+    def init_display(self):
+        self.reset()
+        self.cs(0)
+        self.dc(0)
+        # initialise as usual, except with shortest pre-charge periods and highest clock frequency
+        self.spi.write(bytearray([
+            0xae, 0x20,0x00, 0x40, 0xa1, 0xa8,63, 0xc8, 0xd3,54, 0xda,0x12, 0xd5,0xf0, 0xd9,0x11, 0xdb,0x20, 0x81,0x7f,
+            0xa4, 0xa6, 0x8d,0x14, 0xad,0x30, 0xaf]))
```

Comment:
There are a couple of these "magic value" arrays. I'm sure those are just streams of commands for the display controller, and well documented in the datasheet, but it may be nice to reference the source somewhere, or document what the bytes mean?